src/data-types/numeric.ts

```diff
@@ -1,18 +1,19 @@
+import { DataType, ParameterData } from '../data-type';
 const NumericN = require('./numericn');
 
-module.exports = {
+const Numeric:DataType = {
```

Comment:
If you let TypeScript know that `resolvePrecision` and `resolveScale` are always defined on this datatype, you can get rid of all the `!` on `this.resolvePrecision` and `this.resolveScale`:

```suggestion
const Numeric: DataType & { resolveScale: NonNullable<DataType['resolveScale']>, resolvePrecision: NonNullable<DataType['resolvePrecision']> }= {
```

This is not super beautiful, but nicer than sprinkling the `!` in all these places. 🤷‍♂ Similar to `any`, `!` can hide bugs and thus I'd try to reduce it's usage as much as possible.

Reply:
wandering, can 'precision' and 'scale' 'possibly undefined issue' under 'writeTypeInfo', 'writeParameterData' be solved using this similar fashion?